psydac/linalg/identity.py

```diff
@@ -37,28 +60,83 @@ class IdentityMatrix( Matrix, IdentityLinearOperator ):
     # Deferred methods
     #-------------------------------------
     def toarray( self ):
-        return dense_id(*self.shape)
+        if hasattr(self.codomain, 'dtype'):
+            return dense_id(*self.shape, dtype=self.codomain.dtype)
+        else:
+            return dense_id(*self.shape)
 
     def tosparse( self ):
-        return sparse_id(*self.shape)
+        if hasattr(self.codomain, 'dtype'):
+            return sparse_id(*self.shape, dtype=self.codomain.dtype)
+        else:
+            return sparse_id(*self.shape)
+    
+    def copy(self):
+        return IdentityMatrix(self.domain)
+
+    def __neg__(self):
+        raise NotImplementedError()
+
+    def __mul__(self, a):
+        raise NotImplementedError()
+
+    def __rmul__(self, a):
+        raise NotImplementedError()
+
+    def __add__(self, m):
+        raise NotImplementedError()
+
+    def __sub__(self, m):
+        raise NotImplementedError()
+
+    def __imul__(self, a):
+        raise NotImplementedError()
+
+    def __iadd__(self, m):
+        raise NotImplementedError()
+
+    def __isub__(self, m):
+        raise NotImplementedError()
 
 class IdentityStencilMatrix( StencilMatrix ):
-    def __init__(self, V, p=None):
-        assert V.ndim == 1
-        n = V.npts[0]
-        p = p or V.pads[0]
+    def __init__(self, V, pads=None):
+        assert pads is None or len(pads) == V.ndim
 
-        super().__init__(V, V, pads=(p,))
+        super().__init__(V, V, pads=pads)
 
-        self._data[:, p] = 1.
+        idslice = tuple([slice(None) for _ in range(V.ndim)] + list(self.pads))
```

Comment:
Very minor comment: I think that this can be simply written as
```python
idslice = (slice(None),) * V.ndim + (*self.pads)
```
or
```python
idslice = (*((slice(None),) * V.ndim), *self.pads)
```
But if you want to avoid creating temporary lists or tuples, then using a generator is the way to go:
```python
idslice = (*(slice(None) for _ in range(V.ndim)), *self.pads)
```

On my computer the second option is the faster.

Reply:
Thanks for the hint!